src/Commands/Server/Tools/AzureProxyTool.cs

```diff
@@ -0,0 +1,511 @@
+            {
+                intent = intentElem.GetString();
+            }
+            if (args.TryGetValue("learn", out var learnElem) && learnElem.ValueKind == JsonValueKind.True)
+            {
+                learn = true;
+            }
+            if (args.TryGetValue("tool", out var toolElem) && toolElem.ValueKind == JsonValueKind.String)
+            {
+                tool = toolElem.GetString();
+            }
+            if (args.TryGetValue("command", out var commandElem) && commandElem.ValueKind == JsonValueKind.String)
+            {
+                command = commandElem.GetString();
+            }
+        }
+
+        if (!string.IsNullOrEmpty(intent) && string.IsNullOrEmpty(tool) && string.IsNullOrEmpty(command) && !learn)
+        {
+            learn = true;
+        }
+
+        if (learn && string.IsNullOrEmpty(tool) && string.IsNullOrEmpty(command))
+        {
+            return await RootLearnModeAsync(request, intent ?? "", cancellationToken);
+        }
+        else if (learn && !string.IsNullOrEmpty(tool) && string.IsNullOrEmpty(command))
+        {
+            return await ToolLearnModeAsync(request, intent ?? "", tool!, cancellationToken);
+        }
+        else if (!learn && !string.IsNullOrEmpty(tool) && !string.IsNullOrEmpty(command))
+        {
+            var toolParams = GetParametersDictionary(args);
+            return await CommandModeAsync(request, intent ?? "", tool!, command!, toolParams, cancellationToken);
+        }
+
+        return new CallToolResponse
+        {
+            Content =
+            [
+                new Content {
+                    Type = "text",
+                    Text = """
+                        The "tool" and "command" parameters are required when not learning
+                        Run again with the "learn" argument to get a list of available tools and their parameters.
+                        To learn about a specific tool, use the "tool" argument with the name of the tool.
+                    """
+                }
+            ]
+        };
+    }
+
+    private string GetRootToolsJson()
+    {
+        if (_cachedRootToolsJson != null)
+        {
+            return _cachedRootToolsJson;
+        }
+
+        var providerMetadataList = _mcpClientProvider.ListProviderMetadata();
+        var tools = new List<Tool>();
+        foreach (var meta in providerMetadataList)
+        {
+            tools.Add(new Tool
+            {
+                Name = meta.Id,
+                Description = meta.Description,
+            });
+        }
+        var toolsResult = new ListToolsResult { Tools = tools };
+        var toolsJson = JsonSerializer.Serialize(toolsResult, AzureProxyToolSerializationContext.Default.ListToolsResult);
+        _cachedRootToolsJson = toolsJson;
+
+        return toolsJson;
+    }
+
+    private async Task<string> GetToolListJsonAsync(RequestContext<CallToolRequestParams> request, string tool)
+    {
```

Comment:
A C# specific question, since we do caching, will the async-await engine takes care of the memory visibility of these member caches?